src/object/cli_obj.c

```diff
@@ -1785,7 +1785,7 @@ obj_shard_task_sched(struct obj_auxi_args *obj_auxi, uint64_t epoch)
 	 * the IO involved shards' targets not changed. No any shard task
 	 * re-scheduled for this case, can complete the obj IO task.
 	 */
-	if (obj_auxi->shard_task_scheded == 0)
+	if (obj_auxi->shard_task_scheded == 0 && obj_auxi->obj_task)
```

Comment:
IMHO, obj_task should always be there.  Do you have the daos log available? We better figure out why it is NULL.

Reply:
I agree, but im not seeing that's the case.
i don't have the logs, but you can just apply this patch and revert back this line to reproduce with daos_test -F

note this is intermittent so might have to run a few times to reporoduce. ill try and gather logs meanwhile if that helps

Reply:
client and server log on boro:
/home/mschaara/logs_daos/DAOS-3872

note this has multiple runs in order to reproduce

Reply:
agreed with Di that may need to find what's the problem.

I just tried to apply this patch, and revert L1788, "daos_test -F" works fine for about 10 times running. (with 4~5 servers).

If you rebase on latest master (need to update cart/mercury), still with the segfault? If so can you keep the core file? thx

Reply:
BTW, for your test, did it get RPC timeout? (with "export DD_STDERR=err"   can see timeout msg on stdout)

Reply:
yes i didn't want to land this patch, at least till i know why the task becomes null there.
it was over sockets. no timeout happens.
i did update to latest master yesterday and when i did it was harder to reproduce, but still happened. i will try and investigate what happens to that task today.

Reply:
I just updated to master, and i reproduce now on every other try.
Im looking at this now, so the child shard task can be completed by a different thread between the time it was scheduled in obj_shard_task_sched() in shard_task_sched and hence it can complete the parent API task before we check it and complete it there?

Reply:
i updated the patch and added a sleep before the check. now i can reproduce every-time i run.

Reply:
note that you can now also see the assertion in the CI run:
https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-1545/2/artifact/Functional/job-2019-12-10T17.07-bd36e89-daos_test-daos_core_test/job.log

Reply:
> Im looking at this now, so the child shard task can be completed by a different thread between the time it was scheduled in obj_shard_task_sched() in shard_task_sched and hence it can complete the parent API task before we check it and complete it there?

I reproduced it as well, I think you are right the above case is possible and it is the root cause. When we check the obj_auxi there, the task possible already be completed/destroyed by other threads. I'll push a patch for it.

Reply:
https://github.com/daos-stack/daos/pull/1551 is to fix the assertion/segfault.